### ISSUE TITLE
No single fsync in repository

### ISSUE THREAD

**Issue by funny-falcon** [Bazist/DniproServer#8]
there is no durability without fsync.

**Comment by Bazist** [comment]
First of all, thanks for question.
I am using fflush method (as I know it's wrapper around fsync)
https://github.com/Bazist/DniproServer/search?utf8=%E2%9C%93&q=flush&type=Code

But it's not using in all cases. As was declared in documentation, this database supports ACID and three types of transactions.
So after Begin Transaction it's not necessary call fsync immediately, but on Commit Transaction - yes, it should be.

To test Durability I made test with power off in the middle of writing on disc. After that database file wasn't corrupted and not finished transactions were removed from database file.

**Comment by funny-falcon** [comment]
no, `fflush` is not wrapper around `fsync`. It is just `write` of not already written part of buffer.

> After that database file wasn't corrupted and not finished transactions were removed from database file.

You were lucky. It is possible to make database that will not be corrupted without fsync (by ignoring tail with wrong checksum). But impossible to make sure that "committed" transaction are not removed - that is "Durability" in ACID.

Depending on OS parameters, there could be hours of un-fsynced data in a file cache that didn't reach disk's surface.

**Comment by Bazist** [comment]
Thanks a lot, I will modify the code.
I didn't know about difference under Linux

PS: 
http://stackoverflow.com/questions/2340610/difference-between-fflush-and-fsync
